### ISSUE TITLE
Rule Change: Update no-inner-declaration and remove from eslint:recommended

### ISSUE THREAD

**Issue by Trott** [eslint/eslint#15576]
### ESLint version

8.8.0

### What problem do you want to solve?

The documentation for this rule states that the "function declaration portion rule will be rendered obsolete when block-scoped functions land in ES6". This is referring to what MDN and caniuse call [_Block-level functions_](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions#block-level_functions). They are [very widely supported](https://caniuse.com/mdn-javascript_functions_block_level_functions).

That part of the rule is the only part that's enabled by default.

### What do you think is the correct solution?

Remove `no-inner-declarations` from the recommended rule set.

### Participation

- [X] I am willing to submit a pull request for this change.

**Comment by mdjermanovic** [comment]
I agree that disallowing block-scoped function declarations looks like just a stylistic preference, and therefore it doesn't seem it should be in `eslint:recommended`.

However, function declarations are guaranteed to be block-scoped only in strict code:

```js
"use strict";

const x = 42;

function foo() {
    console.log(x); // 42
    if (true) {
        function x() {}
    }
    console.log(x); // 42
}

foo();
```

In most engines, [`B.3.2 Block-Level Function Declarations Web Legacy Compatibility Semantics`](https://tc39.es/ecma262/#sec-block-level-function-declarations-web-legacy-compatibility-semantics) apply, so the behavior in non-strict code is mostly like that with `var` declarations:

```js
// not strict

const x = 42;

function foo() {
    console.log(x); // undefined (the outer `x` is already shadowed)
    if (true) {
        function x() {}
    }
    console.log(x); // ƒ x() {}
}

foo();
```

This is a legacy feature that should be avoided, as noted in the same MDN article ([Block-level functions in non-strict code](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions#block-level_functions_in_non-strict_code)). This seems like something that should be in `eslint:recommended`.

Would it be better to keep the rule in `eslint:recommended`, but redesign its options, and set default values so that by default it reports only inner (block-level) function declarations in non-strict code?

**Comment by nzakas** [comment]
> Would it be better to keep the rule in eslint:recommended, but redesign its options, and set default values so that by default it reports only inner (block-level) function declarations in non-strict code?

Im not sure changing how the rule works at this point makes much sense. This isn’t just a difference between strict and non-strict code, it’s also a difference between ES5 strict mode and ES6 strict mode. I suppose we could change the rule to check if the function is in a block scope…but again, I’m just not sure there’s much value in changing the rule at this point.

We could consider removing it from the recommended config in a major version.

I don’t feel very strongly either way, though, so happy to defer to consensus here.

**Comment by nzakas** [comment]
@eslint/eslint-tsc can we come to a conclusion here? I’m in favor of not making any changes at this point

**Comment by mdjermanovic** [comment]
Per our [semver policies](https://github.com/eslint/eslint#semantic-versioning-policy), we can change the default behavior of a rule in a non-major release only if the change fixes a bug or adds support for a newly-added language feature. I think this doesn't fall under those categories, so if we decide on that change, it would have to wait for the next major release.

On the other hand, our semver policies do allow removing rules from `eslint:recommended` in a minor release. However, I think we should be applying that policy only in exceptional cases where the configuration as-is may be unusable for most users, such as the only case in which it was applied so far: `require-atomic-updates` was removed from `eslint:recommended` in a minor version due to a high number of false positives that didn't seem fixable at the time.

I'm also in favor of not making any changes at this point (in a non-major release), but I'm in favor of making a change in the next major release.

**Comment by mdjermanovic** [comment]
As far as I know, this is how specifications and browsers treat block-level function declarations:

* ES3 (there's only non-strict mode)
  * Specification: doesn't specify behavior.
  * Browsers: allowed, but inconsistent behavior between different browsers.
* ES5 non-strict mode
  * Specification: doesn't specify behavior, recommends disallowing or issuing a warning.
  * Browsers: allowed, but inconsistent behavior between different browsers.
* ES5 strict mode
  * Specification: doesn't specify behavior, recommends disallowing or issuing a warning.
  * Browsers: mostly disallowed (SyntaxError)
* ES6+ non-strict mode 
  * Specification: allowed. Per the main spec, they are block-scoped. But, if the host is a browser, then Annex [B.3.2](https://tc39.es/ecma262/#sec-block-level-function-declarations-web-legacy-compatibility-semantics) must apply and they're not block-scoped.
  * Browsers:  Per [B.3.2](https://tc39.es/ecma262/#sec-block-level-function-declarations-web-legacy-compatibility-semantics), allowed but not block-scoped.
* ES6+ strict mode 
  * Specification: allowed, block-scoped.
  * Browsers: allowed, block-scoped.

I thought that Espree/Acorn throws parsing error in ES5 strict mode, but it doesn't. There was an issue https://github.com/acornjs/acorn/issues/118, but it was closed without a conclusion.

I think there's a value in reporting block-level functions in:

* non-strict code regardless of `ecmaVersion`, because they have a confusing legacy behavior.
* `ecmaVersion: 5` code regardless of strict, because it's either the same confusing legacy behavior or a SyntaxError in most ES5 engines.

**Comment by nzakas** [comment]
**TSC Summary:** This issue seeks to either remove `no-inner-declaration` from `eslint:recommended` or change the default behavior of `no-inner-declaration` based on how ES6+ works with inner declarations.

**TSC Question:** Do we want to make any change now? Do we want to make any changes in the next major release?

**Comment by btmills** [comment]
We discussed this in the most recent TSC meeting, and there's appetite for making this rule's behavior more consistent with modern JS usage. nzakas suggested a `legacy: true` option, but that's only an idea. We intentionally left discussion around any implementation for this issue. Whatever form it takes, we can add the option now, but we'd wait for a major release to change the option's default and remove the rule from `eslint: recommended`.

**Comment by nzakas** [comment]
There are two tasks in this issue:

- [ ] Remove `no-inner-declarations` from `eslint:recommended`
- [ ] Add a new option to `no-inner-declarations`

**Comment by Tanujkanti4441** [comment]
Is this issue accepting the PR right now

**Comment by nzakas** [comment]
It looks like @mdjermanovic was planning on doing it, but I'm willing to bet he would be happy to have you do it.
